### ISSUE TITLE
[21.02.1] Pairing with WPS let device connect only once and then never again

### ISSUE THREAD

**Issue by fractalwalkr** [openwrt/openwrt#9369]
Name the tree/revision/version: 
* openwrt-21.02.1-ath79-generic-tplink_archer-c7-v5-squashfs-sysupgrade

Name the affected device: 
* TP Link Archer C7 v5

What does it do that it should not do / what does it not do that it should do:
* Devices which was paired with WPS will only connect once to the AP. After a network restart, it won't connect anymore.

Steps to reproduce:
* Setup WIFI AP:
    Mode = N
    Width = 20Mhz
    WPA2-PSK
    Force CCMP   
* Install WPS via CLI:
  opkg update
  opkg remove wpad-mini
  opkg remove wpad-basic-wolfssl
  opkg install wpad hostapd-utils
  -> OK
* Enable WPS for Wifi / AP
* Pair device via WPS
  hostapd_cli -i wlan1 wps_pbc
  hostapd_cli wps_get_status
  -> OK
* See device connecting to AP and it can be pinged
* Restart network or reboot device
  reboot

What it does:
* Device does not pair again anymore
* You see it only short under "Associated Stations" and then it disconnect again. Device tries over and over and it fails always
* Log says with each attempt: 
 daemon.notice hostapd: wlan1: AP-STA-POSSIBLE-PSK-MISMATCH xx:xx:xx:xx..

What it should do:
* Device should always connect to AP after a successful WPS pairing

What you have already done to workaround/fix the problem
* Downgraded to 19.07.8. No issue there. WPS pairing works with described procedure and the device can always connect again afterwards. So it is not a hardware issue.

**Comment by ptpt52** [comment]
wpa key sync by wps is temporary so it would not work after router reboot.

**Comment by fractalwalkr** [comment]
It's not the WPS which does not work afterwards. It is the pairing of the device to the access point which does not persist.

**Comment by Ansuel** [comment]
Can confirm this. I honestly think the algo is not correctly communicated with WPS.

We should check the changes to hostpad about the WPS part...

**Comment by dunglx** [comment]
> Can confirm this. I honestly think the algo is not correctly communicated with WPS.
> 
> We should check the changes to hostpad about the WPS part...

I meet the same problem.
@Ansuel Please help us check the changes. Thanks!

**Comment by dunglx** [comment]
I see in the source file openwrt/package/network/services/hostapd/file/hostapd.sh, then wpa_psk_file is /var/run/hostapd-$ifname.psk. It is just a temporary file and will be a new file after the router reboot. 
I changed the path to /etc/hostapd-$ifname.psk and the connection by WPS can be keep after router reboot.

**Comment by dunglx** [comment]
It's ok when pairing with one tablet via WPS (both router reboots and tablet reboots.
But, when I pair with a printer, then it's ok when router reboots and not OK when the printer reboot. 
Can anyone give me some ideas for this?

**Comment by Salamandar** [comment]
I can confirm @dunglx 's problem.

**Comment by lukasz1992** [comment]
I wonder if setting ap_setup_locked=1 within wifi iface would help?

**Comment by OlafTitz** [comment]
After some experimenting I found out that what you need to do is:
1. Configure wpa_psk_file to a file in /etc
2. Create that file (empty) and make it writable for group "network"
3. Put that file's path in /etc/sysupgrade.conf to have it backed up
Then you will be able to use manually set per-MAC passphrases too.

The standard configuration should probably include this from the start.

**Comment by ssdnvv** [comment]
> After some experimenting I found out that what you need to do is:

This works great and solves the issue - thanks alot!

**Comment by bitrot-alpha** [comment]
Step-by-step version of @OlafTitz 's solution (you can use text editor of your choosing instead of `vi`):

1. Back up your config via LuCI (System -> Backup/Flash Firmware -> Backup heading -> "Generate Archive" button
2. Log into your router's SSH
3. Open the .sh file with vi editor: `vi /lib/netifd/hostapd.sh`
4. `/wpa_psk_file` to search for `wpa_psk_file`
5. Press N to go to next result. Line should be midway through the file, starts with `[ -z "$wpa_psk_file" ] && set_default wpa_psk_file`
6. Move vi's cursor with `h < j \/ k /\ l >` or arrow keys to the start of `/var/run...`
7. use x key or `d $` shortcut to delete the current path for wpa_psk_file
8. enter new path with `i` (insert) and then type new file path. You can paste with `Ctrl+Shift+V` on Linux or right-click on most modern ssh clients
9. hit Esc key to exit "Insert" mode. Type `:wq` to save and exit.
10. `cd /var/run`
11. Copy the *.psk files to /etc: `cp hostapd*.psk /etc/`
12. `cd /etc`
13. Make `network` the owner of the *.psk files: `chown network:network hostapd*.psk`
14. Verify `chown` worked: `ls -l /etc | grep psk` should say network in the 3rd and 4th columns
15. Open LuCI again (or use vi again to open file in next step) and go to "Backup / Flash Firmware" page again. Switch to "Config" tab. Add the paths of the *.psk files.
16. Hit the green "Save" button to add the paths to `/etc/sysupgrade.conf`
17. Reboot your router and test WPS

I haven't tested this with `uci`, but looking at the `hostapd.sh` file, it will use the `wpa_psk_file` environment variable? if it is already set. I imagine instead of following what I did you could set `wpa_psk_file` inside of `/etc/config/wireless` or using `uci`.

I've wrote this because I was scratching my head on how to do it and this issue still hasn't been fixed on OpenWRT `24.10`!

EDIT: I tried adding `option wpa_psk_file /etc/hostapd[...].psk` under `/etc/config/wireless` section `config wifi-iface [RADIO]` and it worked.
